File: Program's_Contributed_By_Contributors/C++ Programs/games/Sudoku.cpp
#include <iostream>
#define N 9
using namespace std;
int arr[N][N] = {
    {3, 0, 6, 5, 0, 8, 4, 0, 0},
    {5, 2, 0, 0, 0, 0, 0, 0, 0},
    {0, 8, 7, 0, 0, 0, 0, 3, 1},
    {0, 0, 3, 0, 1, 0, 0, 8, 0},
    {9, 0, 0, 8, 6, 3, 0, 0, 5},
    {0, 5, 0, 0, 9, 0, 6, 0, 0},
    {1, 3, 0, 0, 0, 0, 2, 5, 0},
    {0, 0, 0, 0, 0, 0, 0, 7, 4},
    {0, 0, 5, 2, 0, 6, 3, 0, 0}};

// Check whether num exists in Column
bool isPresentInCol(int col, int num)
{
    for (int row = 0; row < N; row++)
        if (arr[row][col] == num)
            return true;
    return false;
}

// Check whether num exists in Row
bool isPresentInRow(int row, int num)
{
    for (int col = 0; col < N; col++)
        if (arr[row][col] == num)
            return true;
    return false;
}

// check num is present in 3x3 box or not
bool isPresentInBox(int InitialRow, int InitialCol, int num)
{

    for (int row = 0; row < 3; row++)
        for (int col = 0; col < 3; col++)
            if (arr[row + InitialRow][col + InitialCol] == num)
                return true;
    return false;
}

// Prints Sudoku
void sudokuGrid()
{
    for (int row = 0; row < N; row++)
    {
        for (int col = 0; col < N; col++)
        {
            if (col == 3 || col == 6)
                cout << " | ";
            cout << arr[row][col] << " ";
        }
        if (row == 2 || row == 5)
        {
            cout << endl;
            for (int i = 0; i < N; i++)
                cout << "---";
        }
        cout << endl;
    }
}

// Empty location and update row and column
bool EmptyGrid(int &row, int &col)
{
    for (row = 0; row < N; row++)
        for (col = 0; col < N; col++)
            if (arr[row][col] == 0)
                return true;
    return false;
}
// Item Not in Box
bool isValidPlace(int row, int col, int num)
{
    return !isPresentInRow(row, num) && !isPresentInCol(col, num) && !isPresentInBox(row - row % 3, col - col % 3, num);
}
bool solveSudoku()
{
    int row, col;
    if (!EmptyGrid(row, col))
        return true;
    for (int num = 1; num <= 9; num++)
    {
        if (isValidPlace(row, col, num))
        {
            arr[row][col] = num;
            if (solveSudoku()) // recursively go for other rooms in the grid
                return true;
            arr[row][col] = 0;
        }
    }
    return false;
}
int main()
{
    if (solveSudoku() == true)
        sudokuGrid();
    else
        cout << "No solution exists";
}

File: Program's_Contributed_By_Contributors/Python_Programs/sudoku.py
def print_grid(grid):
    for row in grid:
        print(" ".join(map(str, row))

def is_valid_move(grid, row, col, num):
    # Check if the number is not in the current row and column
    if num in grid[row] or num in [grid[i][col] for i in range(9)]:
        return False

 
    subgrid_row, subgrid_col = 3 * (row // 3), 3 * (col // 3)
    for i in range(3):
        for j in range(3):
            if grid[subgrid_row + i][subgrid_col + j] == num:
                return False

    return True

def is_valid_sudoku(grid):
    # Check if the initial Sudoku grid is valid
    for row in range(9):
        for col in range(9):
            if grid[row][col] != 0:
                num = grid[row][col]
                grid[row][col] = 0
                if not is_valid_move(grid, row, col, num):
                    return False
                grid[row][col] = num
    return True

def solve_sudoku(grid):
    for row in range(9):
        for col in range(9):
            if grid[row][col] == 0:
                for num in range(1, 10):
                    if is_valid_move(grid, row, col, num):
                        grid[row][col] = num
                        if solve_sudoku(grid):
                            return True
                        grid[row][col] = 0
                return False
    return True

def generate_sudoku():
    
    pass

if __name__ == "__main__":
    sudoku_grid = [
        [5, 3, 0, 0, 7, 0, 0, 0, 0],
        [6, 0, 0, 1, 9, 5, 0, 0, 0],
        [0, 9, 8, 0, 0, 0, 0, 6, 0],
        [8, 0, 0, 0, 6, 0, 0, 0, 3],
        [4, 0, 0, 8, 0, 3, 0, 0, 1],
        [7, 0, 0, 0, 2, 0, 0, 0, 6],
        [0, 6, 0, 0, 0, 0, 2, 8, 0],
        [0, 0, 0, 4, 1, 9, 0, 0, 5],
        [0, 0, 0, 0, 8, 0, 0, 7, 9]
    ]

    if not is_valid_sudoku(sudoku_grid):
        print("Invalid Sudoku puzzle.")
    else:
        if solve_sudoku(sudoku_grid):
            print("Sudoku solution:")
            print_grid(sudoku_grid)
        else:
            print("No solution exists.")
